### PR TITLE
Unifies Find scripts in PCLConfig

### PR DESCRIPTION
**Pull request by fran6co**
Supersedes https://github.com/PointCloudLibrary/pcl/pull/1416

### CODE REVIEW

**Comment by jspricke** [comment]
Looks really awesome. I can't test it at the moment, but otherwise +1 :).

**Comment by fran6co** [comment]
@ahundt could you check if this pr solves your problem?

**Comment by jspricke** [comment]
I've been talking to a friend (@v4hn) about it and he proposed to do it rather like this:
https://github.com/v4hn/Las-Vegas-Reconstruction/commit/787b6f7dd7e80348f26ce206c6ab343cae3e6881

**Comment by v4hn** [comment]
The main idea here is to provide Find-scripts with the library, which are not provided by cmake or the specific dependencies (yet).
In case at some point a Find script becomes available upstream (e.g. FindEigen3.cmake which is provided starting with cmake 3.0) the only thing required to switch to the upstream version is to remove the Find script from the libraries build - no need to touch libConfig.cmake at all.
Also the generated libConfig.cmake probably looks a bit better...

**Comment by taketwo** [comment]
The idea of eliminating redundancy is appealing. However, I have serious concerns about the proposed implementation. There is always a difference between the CMake code in `Find***.cmake` script and corresponding macro in `PCLConfig.cmake`. As a result, the diff between the config script generated with current master and with your PR is quite substantial. What increases my concern is that in many cases I just don't know why the scripts are different. Mistake in copy-paste? Sloppy coding? Or a workaround for an issue on some platform?

Just to give an example. The code for detection of DepthSense SDK has slight differences between the "find" script and the "config" script. Originally when I contributed it, it was the same, but then someone showed up (actually @v4hn) and pointed that he gets configuration errors. So we introduced changes in #1337.

With such cases in mind, applying your changes feels like opening a can of worms for me. Especially before 1.8.0 release.

**Comment by v4hn** [comment]
On Fri, Nov 13, 2015 at 10:00:09AM -0800, Sergey Alexandrov wrote:
> The idea of eliminating redundancy is appealing. However, I have serious concerns about the proposed implementation. There is always a difference between the CMake code in `Find***.cmake` script and corresponding macro in `PCLConfig.cmake`. As a result, the diff between the config script generated with current master and with your PR is quite substantial. What increases my concern is that in many cases I just don't know why the scripts are different. Mistake in copy-paste? Sloppy coding? Or a workaround for an issue on some platform?

There are differences, yes. But there shouldn't be any.
If it's a workaround for some specific platform, this workaround should be in both parts of the find logic.

> Just to give an example [...] in #1337.

I never noticed the issue number :)

Concerning this particular issue: actually they were not the same before that commit.
the logic in cmake/Modules/FindDSSDK.cmake does not set DSSDK_INCLUDE_DIRS if the module is not found.
The logic in LVRConfig.cmake did this though.
So the bug would not even have appeared without the "duplication". :)

> With such cases in mind, applying your changes feels like opening a can of worms for me. Especially before 1.8.0 release.

imho this is a can of worms that *should* be opened and cleaned up before the next major release.
As there will be a release candidate for the major release,
this is a great opportunity to make sure the resulting "clean" logic is correct for all relevant platforms.

**Comment by taketwo** [comment]
You are probably right about how the things should be. And that this "can" should be opened. However, I would strongly disagree that now is the right time. Granted, we will introduce configuration issues on some systems. What should be the time frame for waiting for people to catch and report them? How much should we postpone the release?

I'd rather release first and then venture into letting the worms out.

**Comment by jspricke** [comment]
* Sergey Alexandrov <notifications@github.com> [2015-11-13 10:36]:
> I'd rather release first and then venture into letting the worms out.

+1

**Comment by fran6co** [comment]
In that case maybe it's better to merge https://github.com/PointCloudLibrary/pcl/pull/1416, release and then merge this PR. I'm going to change it to @v4hn approach.

**Comment by fran6co** [comment]
Rebased and using @v4hn approach. Should be a bit cleaner.

**Comment by fran6co** [comment]
Would be nice to get rid of the ```find_``` macros and simplify ```find_external_library``` a bit, but I think that would need a bit more work.

**Comment by SergioRAgostinho** [comment]
Hey @fran6co. Can you address the merge conflicts once we release 1.8.1? It would be good to integrate this in the early stage of our preparation for 1.9.0.

File: PCLConfig.cmake.in [hunk 1]
```diff
@@ -896,8 +557,6 @@ endif(NOT "${PCL_DEFINITIONS}" STREQUAL "")
 
 pcl_remove_duplicate_libraries(PCL_LIBRARIES PCL_DEDUP_LIBRARIES)
 set(PCL_LIBRARIES ${PCL_DEDUP_LIBRARIES})
-# Add 3rd party libraries, as user code might include our .HPP implementations
-list(APPEND PCL_LIBRARIES ${BOOST_LIBRARIES} ${QHULL_LIBRARIES} ${OPENNI_LIBRARIES} ${OPENNI2_LIBRARIES} ${ENSENSO_LIBRARIES} ${davidSDK_LIBRARIES} ${DSSDK_LIBRARIES} ${RSSDK_LIBRARIES} ${FLANN_LIBRARIES} ${VTK_LIBRARIES})
```
Comment (SergioRAgostinho) [inline PCLConfig.cmake.in]: Is this not necessary?

**Comment by fran6co** [comment]
@SergioRAgostinho I'll rebase it during this week.

**Comment by SergioRAgostinho** [comment]
If possible I would wait 'till the actual 1.8.1 release. In theory there's only one more PR in 1.8.1 which touches CMake related things and it should not generate a conflict with this one, but just wanting to prevent the extra effort. 

Don't forget about my comment/question in line 900.

File: PCLConfig.cmake.in (hunk 1 above)
Comment (fran6co) [inline PCLConfig.cmake.in]: I can't quite remember why I removed it, I'm putting it back just in case.

**Comment by fran6co** [comment]
@SergioRAgostinho rebased.

**Comment by SergioRAgostinho** [comment]
👍  To be merged once the probation period for 1.8.1 is over.

**Comment by SergioRAgostinho** [comment]
@fran6co  Can you rebase things again? I'm gonna merge it and cross my fingers :)

**Comment by ahundt** [comment]
without any testing I looked at the changes and it appears reasonable to me. For some reason I lost track of this.

**Comment by fran6co** [comment]
@SergioRAgostinho done

**Comment by SergioRAgostinho** [comment]
Pandora's box opened 😅

**Comment by ahundt** [comment]
hey only 2 years to get it merged. :-) at least it is done!

**Comment by UnaNancyOwen** [comment]
@SergioRAgostinho Does this PCLConfig.cmake work correctly on Linux and Mac OS?
I think it will work not correctly on Windows.

**Pull request merged**